### ISSUE TITLE
layout sider will render twice on first load

### ISSUE THREAD

**Issue by jiacai2050** [ant-design/ant-design#8983]
### Version
antd 3.1.0 
react 16.2.0

### Environment
chrome

### Reproduction link

[![Edit n56noj18oj](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/n56noj18oj)
```
class MySider extends Component {
  render() {
    console.log('sider')
    return <Sider>Sider</Sider>
  }
}

class App extends Component {
  render() {
    return (
      <Layout>
      <Header>Header</Header>
      <Layout>
        <MySider></MySider>
        <Content>Content</Content>
      </Layout>
      <Footer>Footer</Footer>
    </Layout>



    );
  }
}
```

### Steps to reproduce
When first load, Snippets above will print `sider` twice

### What is expected?

print `sider` only once

### What is actually happening?
When first load, Snippets above will print `sider` twice

**Comment by yesmeck** [comment]
Duplicate of #8347

**Comment by jiacai2050** [comment]
I have read https://github.com/ant-design/ant-design/pull/8077, but it still confuses me, why it need render twice? Could you explain a little more, thanks in advance.

**Comment by yesmeck** [comment]
>  use this method to detect Sider ant-layout-has-sider class is added to layout only after rerendering because we register Sider to Layout by adding it to state.

**Comment by jiacai2050** [comment]
> because we register Sider to Layout by adding it to state.

I can't really understand this. If this action will cause rerender, why can't implement Sider in other way, which don't cause no rerender. 

For me as an user,  rerender is the last thing I want.

**Comment by yesmeck** [comment]
If you have better solution, welcome pr.
